File: Java/DynamicProgramming/DynamicProgrammingOnSubsequences/322-coin-change/coin-change.java
class Solution {
    public int coinChange(int[] coins, int amount) {
        int prev[] = new int[amount+1];

        for(int T=0;T<=amount;T++){
            if(T % coins[0] == 0) prev[T] = T/coins[0];
            else prev[T] = (int)1e9; 
        }

        for(int index=1 ; index< coins.length; index++){
            int curr[] = new int[amount + 1];
            for(int T=0; T<= amount; T++){

                int notTake = 0 + prev[T];
                int take = (int)1e9;
                if(coins[index] <= T){
                    take = 1 + curr[T - coins[index]];
                }
                curr[T] = Math.min(take, notTake);
            }
            prev= curr;
        }
        
        return (prev[amount] >= (int)1e9) ? -1 : prev[amount];
    }
}

File: Python/DynamicProgramming/DynamicProgrammingOnSubsequences/322-coin-change/coin-change.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        INF = 10**9
        target = amount
        prev = [0]*(target+1)
          
        for T in range(target+1):
            if(T % coins[0] == 0):
                prev[T] = T//coins[0]
            else:
                prev[T] =  INF
        
 
        
        for ind in range(1,len(coins)):
            curr = [0]*(target+1)
            for T in range(target +1):
                not_take = 0 + prev[T]
                take = INF
                
                if coins[ind] <= T :
                    take = 1 + curr[T- coins[ind]]

                curr[T] = min(take, not_take)
            prev = curr
        return prev[target] if (prev[target] != INF) else -1